### PR TITLE
Fix torch.arange traced as constant

### PR DESCRIPTION
**Pull request by lara-hdr**
torch.arange is always traced as a constant which makes it impossible to trace correctly TestModel() from the example below.

class TestModel(torch.nn.Module):
  def forward(self, input):
    return torch.arange(input.shape[0])
input = torch.randn(5,3,2)
print(torch.jit.trace(TestModel(), input).graph)

Currently the trace of TestModel() looks like: 

graph(%self : ClassType<TestModel>,
      %input : Float(5, 3, 2)):
  %11 : int = prim::Constant[value=5]()
  %12 : int = prim::Constant[value=4]()
  %13 : int = prim::Constant[value=0]()
  %14 : Device = prim::Constant[value="cpu"]()
  %15 : bool = prim::Constant[value=0]()
  %16 : Long(5) = aten::arange(%11, %12, %13, %14, %15)
  return (%16)

This PR will allow the trace to have a variable value for %11.
The trace of TestModel() with this PR's modifs looks like:

graph(%self : ClassType<TestModel>,
      %input : Float(5, 3, 2)):
  %2 : int = prim::Constant[value=0]()
  %3 : int = aten::size(%input, %2)
  %4 : Long() = prim::NumToTensor(%3)
  %11 : Scalar = prim::ImplicitTensorToNum(%4)
  %12 : int = prim::Constant[value=4]()
  %13 : int = prim::Constant[value=0]()
  %14 : Device = prim::Constant[value="cpu"]()
  %15 : bool = prim::Constant[value=0]()
  %16 : Long(5) = aten::arange(%11, %12, %13, %14, %15)
  return (%16)

More info : https://github.com/pytorch/pytorch/issues/20075

### CODE REVIEW

**Comment by lara-hdr** [comment]
@eellison @bddppq could you please take a look at this PR?

File: test/test_jit.py [hunk 1]
```diff
@@ -1549,6 +1549,26 @@ def test_trace_size(self):
     def test_trace_size_with_grad(self):
         self.do_trace_size(True)
 
+    def do_trace_arange(self, requires_grad):
+        def fn(x):
+            return torch.arange(x.shape[0])
```
Comment (eellison) [inline test/test_jit.py]: maybe add test for tracing `arange` with a scalar input as well, and another test where both `start` and `end` are specified ?

**Review by eellison — COMMENTED** [review]
LGTM so long as nothing breaks. Since i'm a little unfamiliar with this part of the code base do you mind adding a few more tests ?

Waiting on someone else for the ONNX changes.

**Comment by lara-hdr** [comment]
@eellison, I added 2 more test cases.
@bddppq @BowenBao could you please help reviewing as well?
Trying to get this in asap since it is blocking multiple models.. thanks!

File: test/test_jit.py [hunk 2]
```diff
@@ -1549,6 +1549,40 @@ def test_trace_size(self):
...
+        self.assertEqual(traced_arange_start_end(x), arange_start_end(x))
+
+    def test_trace_arange(self):
+        self.do_trace_arange(False)
+
+    # test the different graph_executor path that happens when
+    # gradients are required and sizes are involved
+    def test_trace_arange_with_grad(self):
+        self.do_trace_arange(True)
+
```
Comment (spandantiwari) [inline test/test_jit.py]: It will be helpful to add an ONNX backend test point for this as well.

**Review by spandantiwari — APPROVED** [review]
The ONNX part looks OK. Please consider adding a backend test point (in `test_pytorch_onnx_onnxruntime.py`) for this. I agree with @eellison that we should ensure that all tests pass with this change.

File: tools/autograd/templates/python_torch_functions.cpp [hunk 3]
```diff
@@ -112,7 +112,7 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
   static PythonArgParser parser({
     "arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
```
[the remark this answers is not in the document]
Comment (lara-hdr) [reply tools/autograd/templates/python_torch_functions.cpp]: we are setting traceable=true on the whole parser, so for the 2 arange signatures.

Here is the signature of PythonArgParser::PythonArgParser(); it takes a vector of string (here the 2 signatures of arange) and a boolean (traceable):
https://github.com/pytorch/pytorch/blob/88e4cee3e70aac95dd2c18b898808ce3426cb3c9/torch/csrc/utils/python_arg_parser.cpp#L580

**Comment by lara-hdr** [comment]
@spandantiwari, ORT test added

**Comment by lara-hdr** [comment]
@houseroad could you take a look at this please?

**Review by houseroad — APPROVED** [review]
Looks good, thanks!

**Pull request closed without merging**